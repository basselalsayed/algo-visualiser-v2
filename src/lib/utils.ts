import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

type T0 =
  | 'string'
  | 'number'
  | 'bigint'
  | 'boolean'
  | 'symbol'
  | 'undefined'
  | 'object'
  | 'function';

export function assert(value: unknown, type: T0): asserts value is T0 {
  const actualType = typeof value;

  if (type === 'object') {
    if (value === null) {
      throw new TypeError(`Expected type '${type}', but received null`);
    }
  }

  if (actualType !== type) {
    throw new TypeError(
      `Expected type '${type}', but received '${actualType}'`
    );
  }
}

export function typeIs(value: unknown, type: T0): value is T0 {
  try {
    assert(value, type);
    return true;
  } catch {
    return false;
  }
}

export function convertToMilliseconds(
  start: DOMHighResTimeStamp,
  end: DOMHighResTimeStamp
) {
  const milliseconds = end - start;
  return Math.floor(milliseconds * 1000) / 1000; // 3 decimal places
}

export function convertToSeconds(
  start: DOMHighResTimeStamp,
  end: DOMHighResTimeStamp
) {
  const seconds = (end - start) / 1000;
  return Math.floor(seconds * 1000) / 1000;
}

export function elementIsFullyScrolled(element: HTMLElement): boolean {
  return (
    Math.floor(element.scrollHeight - element.scrollTop) ===
    Math.floor(element.clientHeight)
  );
}
