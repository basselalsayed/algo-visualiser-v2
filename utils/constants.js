import path from 'node:path';

export const DIRECTORIES = [
  'algorithms',
  'components',
  'contexts',
  'data',
  'hooks',
  'lib',
  'store',
];

export const ROOT_PATH = path.resolve(import.meta.dirname, '..');
