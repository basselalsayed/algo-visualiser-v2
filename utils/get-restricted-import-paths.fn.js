import fs from 'node:fs';
import path from 'node:path';

const __dirname = new URL('.', import.meta.url).pathname;

/**
 * Recursively walks through a directory and collects relative paths to all index.ts or index.tsx files.
 *
 * @param {string} dir - Absolute path to start walking from.
 * @param {string} baseRelativePath - Relative path from the module root (used for nested structure).
 * @returns {string[]} - List of relative paths to directories containing index files.
 */
function findIndexFileDirs(dir, baseRelativePath = '') {
  const indexDirs = [];
  const entries = fs.readdirSync(dir);

  for (const entry of entries) {
    const fullPath = path.join(dir, entry);
    const relPath = path.join(baseRelativePath, entry);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      indexDirs.push(...findIndexFileDirs(fullPath, relPath));
    } else if (
      (entry === 'index.ts' || entry === 'index.tsx') &&
      baseRelativePath !== ''
    ) {
      indexDirs.push(baseRelativePath);
    }
  }

  return indexDirs;
}

/**
 * Generates a list of restricted import paths for all index.ts(x) files in a module.
 *
 * @param {string} dirAbsolutePath - Absolute path to the module directory (e.g. src/components)
 * @param {string} alias - Import alias for the module (e.g. @/components)
 * @returns {Array<{ name: string, message: string }>}
 */
function getRestrictedIndexImports(dirAbsolutePath, alias) {
  const indexDirs = findIndexFileDirs(dirAbsolutePath);

  const staticUpwardPaths = [
    '../../../index',
    '../../..',
    '../../index',
    '../..',
    '../index',
    '..',
  ].map((name) => ({
    message: `Avoid importing '${name}'. Use aliased or direct file paths.`,
    name,
  }));

  const indexPathViolations = indexDirs.flatMap((relPath) => [
    {
      message: `Avoid importing index file: ${alias}/${relPath}/index.ts. Import specific files instead.`,
      name: `${alias}/${relPath}`,
    },
    {
      message: `Avoid importing index file: ../${relPath}/index.ts. Import specific files instead.`,
      name: `../${relPath}`,
    },
  ]);

  return [...staticUpwardPaths, ...indexPathViolations];
}

export { __dirname, getRestrictedIndexImports };
