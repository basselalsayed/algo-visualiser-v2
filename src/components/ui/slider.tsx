import * as SliderPrimitive from '@radix-ui/react-slider';
import * as React from 'react';

import { cn } from '@/lib/utils';

const Slider = ({
  className,
  ref,
  ...props
}: React.ComponentPropsWithRef<typeof SliderPrimitive.Root>) => (
  <SliderPrimitive.Root
    ref={ref}
    className={cn(
      'relative flex w-full touch-none items-center select-none',
      className
    )}
    {...props}
  >
    <SliderPrimitive.Track className='bg_grad_accent/20 relative h-1.5 w-full grow overflow-hidden rounded-full'>
      <SliderPrimitive.Range className='grad-accent absolute h-full' />
    </SliderPrimitive.Track>
    <SliderPrimitive.Thumb className='border-grad-accent-mix bg-background focus-visible:ring-ring block h-4 w-4 rounded-full border shadow-sm transition-colors focus-visible:ring-1 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50' />
  </SliderPrimitive.Root>
);
Slider.displayName = SliderPrimitive.Root.displayName;

export { Slider };
