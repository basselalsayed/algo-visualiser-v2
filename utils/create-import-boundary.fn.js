import path from 'node:path';

import { DIRECTORIES, ROOT_PATH } from './constants.js';
import { getRestrictedIndexImports } from './get-restricted-import-paths.fn.js';

/**
 * @param {string} dir
 */
const getAlias = (dir) => `@/${dir}`;

/**
 * @param {string} dir
 */
export function createImportBoundary(dir) {
  const alias = getAlias(dir);
  const glob = `src/${dir}/**/*.{ts,tsx}`;

  const forbid = DIRECTORIES.flatMap((_dir) =>
    _dir === dir ? undefined : [`${getAlias(_dir)}/**/*`, `${getAlias(_dir)}/*`]
  ).filter(Boolean);

  const absoluteDir = path.resolve(ROOT_PATH, 'src', dir);

  const restrictedIndexPaths = getRestrictedIndexImports(absoluteDir, alias);

  return {
    files: [glob],
    rules: {
      'import/no-internal-modules': [
        'error',
        {
          forbid,
        },
      ],
      'no-restricted-imports': [
        'error',
        {
          paths: [
            ...restrictedIndexPaths,
            {
              message: `Avoid importing '${alias}' from within ${dir}/. Use direct file paths.`,
              name: alias,
            },
          ],
          patterns: [
            {
              group: ['**/index', '**/index.ts', '**/index.tsx'],
              message: `Avoid importing index files directly. Import specific files instead.`,
            },
            {
              group: ['../../**/*'],
              message: `Avoid relative imports outside of this module. Use alias paths instead.`,
            },
          ],
        },
      ],
    },
  };
}
