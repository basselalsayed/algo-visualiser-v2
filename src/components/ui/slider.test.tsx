import { renderFragmentSnapshot } from '@/test/utils';

import { Slider } from './slider';

describe('Slider.component', () => {
  it('looks the same', () => {
    expect(renderFragmentSnapshot(<Slider />)).toMatchSnapshot();
  });
});
