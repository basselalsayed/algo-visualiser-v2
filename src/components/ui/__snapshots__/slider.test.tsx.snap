// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Slider.component > looks the same 1`] = `
<DocumentFragment>
  <span
    aria-disabled="false"
    class="relative flex w-full touch-none items-center select-none"
    data-orientation="horizontal"
    dir="ltr"
    style="--radix-slider-thumb-transform: translateX(-50%);"
  >
    <span
      class="bg_grad_accent/20 relative h-1.5 w-full grow overflow-hidden rounded-full"
      data-orientation="horizontal"
    >
      <span
        class="absolute h-full grad-accent"
        data-orientation="horizontal"
        style="left: 0%; right: 100%;"
      />
    </span>
    <span
      style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(0% + 0px);"
    >
      <span
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="block h-4 w-4 rounded-full border border-grad-accent-mix bg-background shadow-sm transition-colors focus-visible:ring-1 focus-visible:ring-ring focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50"
        data-orientation="horizontal"
        data-radix-collection-item=""
        role="slider"
        style=""
        tabindex="0"
      />
    </span>
  </span>
</DocumentFragment>
`;
